examples/vision/video_transformers.py

```diff
@@ -0,0 +1,408 @@
+            frames = np.concatenate(frames, padding)
+
+        frames = frames[None, ...]
+
+        # Initialize placeholder to store the features of the current video.
+        temp_frame_featutes = np.zeros(
+            shape=(1, MAX_SEQ_LENGTH, NUM_FEATURES), dtype="float32"
+        )
+
+        # Extract features from the frames of the current video.
+        for i, batch in enumerate(frames):
+            video_length = batch.shape[1]
+            length = min(MAX_SEQ_LENGTH, video_length)
+            for j in range(length):
+                if np.mean(batch[j, :]) > 0.0:
+                    temp_frame_featutes[i, j, :] = feature_extractor.predict(
+                        batch[None, j, :]
+                    )
+
+                else:
+                    temp_frame_featutes[i, j, :] = 0.0
+
+        frame_features[idx,] = temp_frame_featutes.squeeze()
+
+    return frame_features, labels
+
+
+"""
+Calling `prepare_all_videos()` on `train_df` and `test_df` takes ~20 minutes time to
+complete execution. This is why, to save time, we will use already prepared NumPy arrays.
+"""
+
+"""shell
+wget -q https://git.io/JZmf4 -O top5_data_prepared.tar.gz
+tar xf top5_data_prepared.tar.gz
+"""
+
+train_data, train_labels = np.load("train_data.npy"), np.load("train_labels.npy")
+test_data, test_labels = np.load("test_data.npy"), np.load("test_labels.npy")
+
+print(f"Frame features in train set: {train_data.shape}")
+
+"""
+## Building the Transformer-based model
+
+We will be building on top of the code shared in 
+[this book chapter](https://livebook.manning.com/book/deep-learning-with-python-second-edition/chapter-11) of 
+[Deep Learning with Python (Second ed.)](https://www.manning.com/books/deep-learning-with-python)
+by François Chollet. 
+
+First, self-attention layers that form the basic blocks of a Transformer are
+order-agnostic. Since, videos are ordered sequences of frames we would need our
+Transformer model to account for that. This is done via **positional encodings**. We will
+simply embed the positions of the frames present inside videos with
+[`layers.Embedding()`](https://keras.io/api/layers/core_layers/embedding). We will then
+add these positional embeddings to the precomputed CNN feature maps. 
+"""
+
+
+class PositionalEmbedding(layers.Layer):
+    def __init__(self, sequence_length, output_dim, **kwargs):
+        super().__init__(**kwargs)
+        self.position_embeddings = layers.Embedding(
+            input_dim=sequence_length, output_dim=output_dim
+        )
+        self.sequence_length = sequence_length
+        self.output_dim = output_dim
+
+    def call(self, inputs):
+        # Our inputs are of shape: `(batch_size, frames, num_features)
+        length = tf.shape(inputs)[1]
+        positions = tf.range(start=0, limit=length, delta=1)
+        embedded_positions = self.position_embeddings(positions)
+        return inputs + embedded_positions
+
+    def compute_mask(self, inputs, mask=None):
```

Comment:
No need, you don't use padding/masking in this model

Reply:
@fchollet I am actually padding the feature vector in case a video has a shorter number of frames than `max_seq_length`.

Reply:
Got it. In this case you should make sure the `TransformerEncoder` is receiving the correct mask argument (correctly it gets None, it seems). I'd encourage to display the masks you're getting in `TransformerEncoder.call()` as a way to make sure you're doing the right thing

Reply:
Is there a separate `mask` tensor we could calculate the same we did in the prequel example and pass that to the PositionalEmbedding layer? Also, a small doubt is if I don't specify the `mask` argument in the `TransformerEncoder` instance will it be set automatically?

Reply:
Your input sequences (that go into the PositionalEmbedding layer) have all-zero locations, right? You can just implement a `compute_mask` that turns them into 0s. Then if you don't manually pass a `mask` argument, the mask generated by `compute_mask` will be passed to the next layer (TransformerEncoder). But make sure to print what you're getting, while you debug. Also, the TransformerEncoder will need to reshape the mask.

Reply:
Got it. It's clear now. It's likely that none of the videos are shorter than `max_seq_length` in this case which is why the following does not ever arise:

> Your input sequences (that go into the PositionalEmbedding layer) have all-zero locations, right?

Also, regarding

> Then if you don't manually pass a mask argument, the mask generated by compute_mask will be passed to the next layer (TransformerEncoder). But make sure to print what you're getting, while you debug. Also, the TransformerEncoder will need to reshape the mask.

The current `compute_mask()` from `PositionalEmbedding` returns a tensor of shape `(batch_size, max_seq_length, num_features)`. As per [the documentation](https://keras.io/api/layers/attention_layers/multi_head_attention/): 

> attention_mask: a boolean mask of shape [B, T, S], that prevents attention to certain positions. The boolean mask specifies which query elements can attend to which key elements, 1 indicates attention and 0 indicates no attention. Broadcasting can happen for the missing batch dimensions and the head dimension.

I guess it (the mask returned from  `PositionalEmbedding`) does not have to be reshaped from  `TransformerEncoder` anymore then? Let me know if I am missing something.

Reply:
> It's likely that none of the videos are shorter than max_seq_length

Then you don't need masking.

Right now you don't use masking, because you pass `mask=None` to the TransformerEncoder. If you don't need masking, cool, but then you can get rid of the `compute_output_mask` method here as well as the `mask` argument in the TransformerEncoder.

Reply:
> Then you don't need masking.

Yes, but I wanted the workflow to be extensible for the masking cases. too. In that case, will the workaround be enough mentioned in my previous comment? Of course, I won't specify the `mask` to be `None` anymore then.

Reply:
Ok, sounds good! Yeah I guess it should work.

Reply:
Thanks. Here's what I will do next:

* Try out the suggestions discussed. If it works right off the bat, I will make the necessary changes and generate the files. 
* If it doesn't, then I will remove the `compute_mask()` function and all the occurrences of `mask` and generate the files. 

Sounds good?

Reply:
I think I am computing the mask in the wrong way. The mask returned by `compute_mask()` has a shape of `(batch_size, max_seq_length, num_features)` which is causing the following:

```python
ValueError: Dimensions must be equal, but are 20 and 1024 for '{{node model/transformer_layer/multi_head_attention/softmax/add}} = AddV2[T=DT_FLOAT](model/transformer_layer/multi_head_attention/einsum/Einsum, model/transformer_layer/multi_head_attention/softmax/mul)' with input shapes: [?,1,20,20], [?,1,20,1024].
```

@fchollet any workaround you would suggest?

Reply:
Your input is a sequence of vectors, and for each vector, either you're masking the vector or you're attending to the vector -- you have a single mask value for the entire vector. So your padding mask has the natural shape `(B, length)`. Now, your MHA layer actually expects mask shape `(B, T, S)` which means `(B, query_length, value_length)` which in our case means `(B, length, length)` (query and value are the same).

You can just do something like:

```
mask = tf.reduce_any(tf.cast(inputs, 'bool'), axis=-1)
mask = mask[:, tf.newaxis, :]  # or mask[:, :, tf.newaxis]
```

The first step computes the boolean sequence step mask, the second step broadcasts to the desired shape. I'd recommend doing the second step inside TransformerEncoder.

Reply:
That makes sense. Thank you. Let me look into incorporating it now.